### DISCUSSION TITLE
Question: What's the long-term viability of writing Rust code targeting this frontend?

### DISCUSSION THREAD

**Discussion by kerkeslager** [Rust-GCC/gccrs#2870] [General]
I've got a few projects written in C. I've basically chosen C, not because I love the language, but because I can be sure that it will continue to be well-supported for the foreseeable future on a variety of platforms. From a language perspective it's clear to me that Rust is more suitable for my goals than C, but my outside impression is that it has only one mature compiler implementation. And while the Apache/MIT license that the mainstream Rust compiler is under is fine, my next project is targeted at the GNU ecosystem and it would be preferable to have an implementation licensed under the GPL.

I'd like to commit to Rust code targeting this compiler frontend for my next project, but before I commit to that, I'd like to get some clarity on the long-term viability of this project. Part of my concern is that there was a previous attempt at a Rust frontend for GCC which fell off my radar for a while and I'm worried that might happen again (maybe it was reborn here, but I don't have the means to track that). Obviously folks working on this project are likely to think it's viable, but more specifically:

* How is this project being funded?
* How many developers are committed to working on this, and for how long?
* Are any of the aforementioned developers working on this project full time, or committed to some number of hours per week (or some other time commitment model I haven't thought of)?
* Is there a timeline for merging into GCC?

Just to be clear, none of this is intended to be criticism, I'm just asking questions because I don't know the answers. I'm genuinely rooting for you to succeed. If I were at all qualified to do so, I'd be contributing to this project because I believe if it succeeds it will solve a real problem I experience. I just want to say in addition to these questions that I appreciate all the hard work y'all have already put into this.

**Comment by CohenArthur** [comment]
Hey there, thanks for your interest in the project :) it's great to see more and more people discover about gccrs and thank you for opening a discussion! Also, sorry we left you unanswered for 4 days - we're a bit drowning in notifications at the moment.

> I've got a few projects written in C. I've basically chosen C, not because I love the language, but because I can be sure that it will continue to be well-supported for the foreseeable future on a variety of platforms. From a language perspective it's clear to me that Rust is more suitable for my goals than C, but my outside impression is that it has only one mature compiler implementation. And while the Apache/MIT license that the mainstream Rust compiler is under is fine, my next project is targeted at the GNU ecosystem and it would be preferable to have an implementation licensed under the GPL.
> 
> I'd like to commit to Rust code targeting this compiler frontend for my next project, but before I commit to that, I'd like to get some clarity on the long-term viability of this project.

Before we go any further, I would like to point out that committing to this compiler frontend for Rust code is way too early. We are still in the "early" development stages of the compiler, and are nowhere close to feature parity with the official Rust compiler. We do not offer all of the stability guarantees that Rust has, and still have a ways to go.

I would suggest checking out [rustc_codegen_gcc](https://github.com/rust-lang/rustc_codegen_gcc). This project is similar in spirit to gccrs, but achieves the goal of compiling Rust code using the GCC project differently. It's much more advanced, complete, and benefits from all the existing safety features of rustc.

> Part of my concern is that there was a previous attempt at a Rust frontend for GCC which fell off my radar for a while and I'm worried that might happen again (maybe it was reborn here, but I don't have the means to track that). Obviously folks working on this project are likely to think it's viable, but more specifically:
> 
>     * How is this project being funded?

This project is being funded by Open Source Security, inc. and Embecosm Ltd. Three people are currently being paid to work on it, @philberty, @P-E-P and myself. We do not work full-time on the project, but can still dedicate a lot of time to it.

>     * How many developers are committed to working on this, and for how long?

At least 3, but we are looking into having more people work on the project as paid contributors. The work will continue being funded until at least a first release of gccrs, and probably for a long time after - we have some very kind, generous and involved people funding us.

>     * Are any of the aforementioned developers working on this project full time, or committed to some number of hours per week (or some other time commitment model I haven't thought of)?

@P-E-P and I are paid to work on this project for 50% of our working time, so that is 20h per week each - in reality, we do a lot more (especially right now as we approach GCC release).
@philberty works on this project for 2 to 3 days a week.

>     * Is there a timeline for merging into GCC?

Yes! We are already merged into GCC, but aren't yet being distributed as part of the main GCC releases. This should change this year with the GCC 14.1 release, around May 2024.

> Just to be clear, none of this is intended to be criticism, I'm just asking questions because I don't know the answers. I'm genuinely rooting for you to succeed. If I were at all qualified to do so, I'd be contributing to this project because I believe if it succeeds it will solve a real problem I experience. I just want to say in addition to these questions that I appreciate all the hard work y'all have already put into this.

Thank you :D

**Reply by kerkeslager** [reply]
Thank you for taking the time to respond to me!

> [S]orry we left you unanswered for 4 days - we're a bit drowning in notifications at the moment.

No worries! I'm further behind on some notification streams than 4 days at the moment.

> Before we go any further, I would like to point out that committing to this compiler frontend for Rust code is way too early. We are still in the "early" development stages of the compiler, and are nowhere close to feature parity with the official Rust compiler. We do not offer all of the stability guarantees that Rust has, and still have a ways to go.

Yep! Understood--my intention was to target the mainline Rust compiler initially. I'm not so concerned with targeting this compiler initially, I just wanted some assurance that I would be able to target the gcc toolchain with my Rust code at some point in the forseeable future, before I commit to writing the project in Rust. Thank you for your candor here, though.

It sounds like I'll be choosing Rust for my project!
